[run-benchmarks] The browser benchmark runner harness should be able to specify MotionMark subtests to run

xw opened the pull request:
#### 653b530b9ffd3f4440fc6636b2da3c011633215f
<pre>
[run-benchmarks] The browser benchmark runner harness should be able to specify MotionMark subtests to run
<a href="https://bugs.webkit.org/show_bug.cgi?id=257663">https://bugs.webkit.org/show_bug.cgi?id=257663</a>
rdar://110187208

Reviewed by Dewei Zhu.

This patch adds support for MotionMark subtests in run-benchmarks, similar to the support already included for JetStream2.

Subtests can be listed with the `--list-subtests` flag, and come in the form {SUITE}/{TEST}. MotionMark&apos;s default run (without any subtests specified) contains just the MotionMark suite in order, but the test plan includes all the other debug suites and tests as well for running separately.

To run subtests, use the --subtests flag and specify a list of subtests (as listed in --list-subtests), separated by spaces. The runner will run only those subtests in each iteration, and report on only those results.

Also, some more under-the hood improvements:
 - Added a new field to the plan, &apos;subtest_entry_point&apos;, that changes the entry point of the test if --subtests is specified.
 - Changes to webserver/motionmark.patch include not auto-loading the benchmark if `developer.html` is loaded in (because the debug runner already starts tests automatically if encoded), and extending the debug runner to accept multiple suites and tests as arguments.

* Tools/Scripts/webkitpy/benchmark_runner/benchmark_runner.py:
(BenchmarkRunner._run_benchmark):
* Tools/Scripts/webkitpy/benchmark_runner/browser_driver/browser_driver.py:
(BrowserDriver.profile): Fixed a bug where using the --profile flag would crash the benchmark harness, and added logging to note when the placeholder context is being called.
* Tools/Scripts/webkitpy/benchmark_runner/data/patches/webserver/MotionMark.patch:
* Tools/Scripts/webkitpy/benchmark_runner/data/plans/motionmark1.2.1.plan:
* Tools/Scripts/webkitpy/benchmark_runner/webserver_benchmark_runner.py:
(WebServerBenchmarkRunner):

Canonical link: <a href="https://commits.webkit.org/264906@main">https://commits.webkit.org/264906@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/15ae7b3021be43c67aeb121b603fda918c1492f3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/9009 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/9296 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/9515 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/10662 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/8972 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/9017 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/11283 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/9264 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/11823 "1 flakes 96 failures") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/9155 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/10130 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/7900 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/10821 "Built successfully") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/9133 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/7425 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/8232 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/15715 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/8534 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/8380 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/11703 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/8880 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/7242 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/8125 "Built successfully") | | | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/2197 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/12340 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/8630 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

dewei-zhu commented on Tools/Scripts/webkitpy/benchmark_runner/browser_driver/browser_driver.py:
```diff
@@ -52,7 +52,7 @@ def prevent_sleep(self, timeout):
     @contextmanager
-    def profile(self, timeout):
+    def profile(self, output_path, profile_filename, timeout=300):
         yield
```
May worth making it warning or error when profile is specifying in argument but only a dummy profiling context is available.

xw replied:
Noted, will add some logging here. I plan to add support for `ktrace artrace` in the near future, so this will hopefully not stay a dummy profiling context for long.

dewei-zhu commented on Tools/Scripts/webkitpy/benchmark_runner/data/patches/webserver/MotionMark.patch:
```diff
@@ -47,30 +114,37 @@ index d45155e..b10635a 100644
-+            }
-+        }
-+        xhr.send(results);
++       console.log("[DEBUG] Finished last iteration")
```
Remove?

dewei-zhu commented on Tools/Scripts/webkitpy/benchmark_runner/data/patches/webserver/MotionMark.patch:
```diff
@@ -47,30 +114,37 @@ index d45155e..b10635a 100644
++       console.log("[DEBUG] Finished last iteration")
++       // submit result to server
++       var results = JSON.stringify(this._computeResultsReports());
++       console.log(results)
```
Ditto.

xw replied:
Updated in revised commit, forgot to take this out beforehand wups

dewei-zhu commented on Tools/Scripts/webkitpy/benchmark_runner/data/plans/motionmark1.2.1.plan:
```diff
@@ -1,11 +1,12 @@
     "count": 1,
-    "local_git_archive": "PerformanceTests/MotionMark@86daf4fcece9d2dafe67ef0de1e23bda72b56378",
-    "github_source": "https://github.com/WebKit/WebKit/tree/86daf4fcece9d2dafe67ef0de1e23bda72b56378/PerformanceTests/MotionMark",
+    "local_git_archive": "PerformanceTests/MotionMark@4b28b98d37306eb5ae9ab8bca504590295f64210",
```
where is this revision coming from?

xw replied:
This revision is the most recent update to the MotionMark harness, from grorg in September 2022.
https://github.com/WebKit/WebKit/commit/4b28b98d37306eb5ae9ab8bca504590295f64210

xw replied:
Per discussion with smfr: we actually want to leave this revision as-is. Will update to reflect this shortly.

dewei-zhu commented on Tools/Scripts/webkitpy/benchmark_runner/benchmark_runner.py:
```diff
@@ -167,7 +170,7 @@ def _run_benchmark(self, count, web_root):
                     else:
-                        raise Exception('Plan does not contain entry_point or test_files')
+                        raise Exception('Plan does not contain entry_point/subtest_entry_point or test_files')
 
```
This doesn't seem to accurately describe the situation.

xw replied:
How so? We hit this line if the plan contains no entry points (either the default entry_point or the new subtest_entry_point), or if no files are specified. Is there a way we can make this clearer?

dewei-zhu replied:
We only hit this else condition when `'entry_point' in self._plan` is false.

dewei-zhu commented on Tools/Scripts/webkitpy/benchmark_runner/webserver_benchmark_runner.py:
```diff
@@ -63,6 +62,8 @@ def _run_one_test(self, web_root, test_file, iteration):
             self._http_server_driver.serve(web_root)
             url = urljoin(self._http_server_driver.base_url(), self._plan_name + '/' + test_file + self._construct_subtest_url(self._subtests))
+            if '?' not in url:
+                url = url.replace('&', '?', 1)
```
Could you tell me more about this change?

xw replied:
This was originally omitted because of an oversight, since the JetStream2 entry point already has a parameter in it (`index.html?report=true`) we can just keep appending with `&` in between the parameters. However, since the MotionMark subtest entry point does not have a parameter already (`developer.html`), we need to replace the first `?` in the query with a `&`. This could have also been done in `_construct_subtest_url()`, but to me here makes more sense in case we ever want to include more parameters before the subtest ones in the future.

dewei-zhu reviewed (approved):
r=me

xw committed:
[run-benchmarks] The browser benchmark runner harness should be able to specify MotionMark subtests to run
https://bugs.webkit.org/show_bug.cgi?id=257663
rdar://110187208

Reviewed by Dewei Zhu.

This patch adds support for MotionMark subtests in run-benchmarks, similar to the support already included for JetStream2.

Subtests can be listed with the `--list-subtests` flag, and come in the form {SUITE}/{TEST}. MotionMark's default run (without any subtests specified) contains just the MotionMark suite in order, but the test plan includes all the other debug suites and tests as well for running separately.

To run subtests, use the --subtests flag and specify a list of subtests (as listed in --list-subtests), separated by spaces. The runner will run only those subtests in each iteration, and report on only those results.

Also, some more under-the hood improvements:
 - Added a new field to the plan, 'subtest_entry_point', that changes the entry point of the test if --subtests is specified.
 - Changes to webserver/motionmark.patch include not auto-loading the benchmark if `developer.html` is loaded in (because the debug runner already starts tests automatically if encoded), and extending the debug runner to accept multiple suites and tests as arguments.

* Tools/Scripts/webkitpy/benchmark_runner/benchmark_runner.py:
(BenchmarkRunner._run_benchmark):
* Tools/Scripts/webkitpy/benchmark_runner/browser_driver/browser_driver.py:
(BrowserDriver.profile): Fixed a bug where using the --profile flag would crash the benchmark harness, and added logging to note when the placeholder context is being called.
* Tools/Scripts/webkitpy/benchmark_runner/data/patches/webserver/MotionMark.patch:
* Tools/Scripts/webkitpy/benchmark_runner/data/plans/motionmark1.2.1.plan:
* Tools/Scripts/webkitpy/benchmark_runner/webserver_benchmark_runner.py:
(WebServerBenchmarkRunner):

Canonical link: https://commits.webkit.org/264906@main
